frontend/apps/reader/modules/readerpagemap.lua

```diff
@@ -0,0 +1,416 @@
+                range = Geom:new{
+                    x = 0, y = 0,
+                    w = Screen:getWidth(),
+                    h = Screen:getHeight(),
+                },
+                direction = BD.flipDirectionIfMirroredUILayout("east")
+            }
+        }
+    }
+
+    self.pagelist_menu = CenterContainer:new{
+        dimen = Screen:getSize(),
+        covers_fullscreen = true, -- hint for UIManager:_repaint()
+        pl_menu,
+    }
+
+    -- buid up menu widget method as closure
+    local pagemap = self
+    function pl_menu:onMenuChoice(item)
+        pagemap.ui.link:addCurrentLocationToStack()
+        pagemap.ui.rolling:onGotoXPointer(item.xpointer)
+    end
+
+    pl_menu.close_callback = function()
+        UIManager:close(self.pagelist_menu)
+    end
+
+    pl_menu.show_parent = self.pagelist_menu
+    self.refresh = function()
+        pl_menu:updateItems()
+    end
+
+    UIManager:show(self.pagelist_menu)
+    return true
+end
+
+function ReaderPageMap:wantsPageLabels()
+    return self.has_pagemap and self.use_page_labels
+end
+
+function ReaderPageMap:getCurrentPageLabel(clean_label)
+    -- Note: in scroll mode with PDF, when multiple pages are shown on
+    -- the screen, the advertized page number is the greatest page number
+    -- among the pages shown (so, the page number of the partial page
+    -- shown at bottom of screen).
+    -- For consistency, getPageMapCurrentPageLabel() returns the last page
+    -- label shown in the view if there are more than one (or the previous
+    -- one if there is none).
+    local label = self.ui.document:getPageMapCurrentPageLabel()
+    return clean_label and self:cleanPageLabel(label) or label
+end
+
+function ReaderPageMap:getFirstPageLabel(clean_label)
+    local label = self.ui.document:getPageMapFirstPageLabel()
+    return clean_label and self:cleanPageLabel(label) or label
+end
+
+function ReaderPageMap:getLastPageLabel(clean_label)
+    local label = self.ui.document:getPageMapLastPageLabel()
+    return clean_label and self:cleanPageLabel(label) or label
+end
+
+function ReaderPageMap:getXPointerPageLabel(xp, clean_label)
+    local label = self.ui.document:getPageMapXPointerPageLabel(xp)
+    return clean_label and self:cleanPageLabel(label) or label
+end
+
+function ReaderPageMap:getRenderedPageNumber(page_label, cleaned)
+    -- Only used from ReaderGoTo. As page_label is a string, no
+    -- way to use a binary search: do a full scan of the PageMap
+    -- here in Lua, even if it's not cheap.
+    local page_list = self.ui.document:getPageMap()
+    for k, v in ipairs(page_list) do
+        local label = cleaned and self:cleanPageLabel(v.label) or v.label
+        if label == page_label then
+            return v.page
+        end
+    end
+end
+
+function ReaderPageMap:addToMainMenu(menu_items)
+    menu_items.page_map = {
+        text = _("Reference pages"),
+        sub_item_table ={
+            {
+                text = _("Reference source info"),
```

Comment:
I think we might want to throw in some `@translators` notes. A reference source just sounds like a dictionary or something.

Reply:
OK, added 2 - feels a bit excessive to add more... Hopefully, translators will meet all of them near each others and will know what it's about.